services/director-v2/src/simcore_service_director_v2/modules/scheduler.py

```diff
@@ -0,0 +1,310 @@
+        # The sidecar only pick up tasks that are in PENDING state
+        comp_tasks_repo: CompTasksRepository = _get_repository(
+            self.db_engine, CompTasksRepository
+        )
+        await comp_tasks_repo.mark_project_tasks_as_pending(project_id, tasks)
+        # notify the sidecar they should start now
+        self.celery_client.send_computation_tasks(
+            user_id=user_id,
+            project_id=project_id,
+            single_tasks=celery_tasks,
+            callback=self._wake_up_scheduler_now,
+        )
+
+    async def _schedule_pipeline(
+        self, user_id: UserID, project_id: ProjectID, iteration: PositiveInt
+    ) -> None:
+        logger.debug(
+            "checking run of project [%s:%s] for user [%s]",
+            project_id,
+            iteration,
+            user_id,
+        )
+
+        pipeline_dag: nx.DiGraph = await self._get_pipeline_dag(project_id)
+        pipeline_tasks: Dict[str, CompTaskAtDB] = await self._get_pipeline_tasks(
+            project_id, pipeline_dag
+        )
+
+        # filter out the tasks with what were already completed
+        pipeline_dag.remove_nodes_from(
+            {
+                node_id
+                for node_id, t in pipeline_tasks.items()
+                if t.state in _COMPLETED_STATES
+            }
+        )
+
+        # update the current status of the run
+        pipeline_result: RunningState = await self._update_run_result(
+            user_id, project_id, iteration, pipeline_tasks
+        )
+
+        if not pipeline_dag.nodes():
+            # there is nothing left, the run is completed, we're done here
+            self.scheduled_pipelines.remove((user_id, project_id, iteration))
+            logger.info(
+                "pipeline %s scheduling completed with result %s",
+                project_id,
+                pipeline_result,
+            )
+            return
+
+        # find the next tasks that should be run now,
+        # this tasks are in PUBLISHED state and all their dependents are completed
+        next_tasks: List[NodeID] = [
+            node_id
+            for node_id, degree in pipeline_dag.in_degree()
+            if degree == 0 and pipeline_tasks[node_id].state == RunningState.PUBLISHED
+        ]
+        if not next_tasks:
+            # nothing to run at the moment
+            return
+
+        # let's schedule the tasks, mark them as PENDING so the sidecar will take them
+        await self._start_tasks(user_id, project_id, pipeline_tasks, next_tasks)
+
+    def _wake_up_scheduler_now(self) -> None:
+        self.wake_up_event.set()
+
+
+async def scheduler_task(scheduler: CeleryScheduler) -> None:
+    while True:
+        try:
+            logger.debug("scheduler task running...")
+            await scheduler.schedule_all_pipelines()
+            with suppress(asyncio.TimeoutError):
+                await asyncio.wait_for(
+                    scheduler.wake_up_event.wait(), timeout=_DEFAULT_TIMEOUT_S
```

Comment:
this is triggered when:
1. a new pipeline has been added
2. a scheduled task completed
3. after 5 seconds (I'm actually considering whether this might still be necessary)

Reply:
i understand. I meant whether we could rename ``wake_up_event`` differently so the line above might read better. 
never mind, not important :-)